src/wayland/data_device/server_dnd_grab.rs

```diff
@@ -1,87 +1,68 @@
-use std::{cell::RefCell, ops::Deref as _, os::unix::io::RawFd, rc::Rc};
-
-use wayland_server::{
-    protocol::{wl_data_device_manager::DndAction, wl_data_offer, wl_pointer, wl_surface},
-    Main,
+use std::{
+    cell::RefCell,
+    sync::{Arc, Mutex},
 };
 
-use crate::{
-    utils::{Logical, Point},
-    wayland::{
-        seat::{AxisFrame, PointerGrab, PointerGrabStartData, PointerInnerHandle, Seat},
-        Serial,
+use wayland_server::{
+    backend::{protocol::Message, ClientId, Handle, ObjectData, ObjectId},
+    protocol::{
+        wl_data_device_manager::DndAction,
+        wl_data_offer::{self, WlDataOffer},
+        wl_surface,
     },
+    DisplayHandle, Resource,
 };
 
-use super::{DataDeviceData, SeatData};
+use crate::wayland::seat::{
+    AxisFrame, ButtonEvent, MotionEvent, PointerGrab, PointerGrabStartData, PointerInnerHandle, Seat,
+};
 
-/// Event generated by the interactions of clients with a server initiated drag'n'drop
-#[derive(Debug)]
-pub enum ServerDndEvent {
-    /// The client chose an action
-    Action(DndAction),
-    /// The DnD resource was dropped by the user
-    ///
-    /// After that, the client can still interact with your resource
-    Dropped,
-    /// The Dnd was cancelled
-    ///
-    /// The client can no longer interact
-    Cancelled,
-    /// The client requested for data to be sent
-    Send {
-        /// The requested mime type
-        mime_type: String,
-        /// The FD to write into
-        fd: RawFd,
-    },
-    /// The client has finished interacting with the resource
-    ///
-    /// This can only happen after the resource was dropped.
-    Finished,
-}
+use super::{DataDeviceHandler, SeatData, ServerDndGrabHandler, SourceMetadata};
 
-pub(crate) struct ServerDnDGrab<C: 'static> {
+pub(crate) struct ServerDnDGrab<D> {
     start_data: PointerGrabStartData,
     metadata: super::SourceMetadata,
     current_focus: Option<wl_surface::WlSurface>,
     pending_offers: Vec<wl_data_offer::WlDataOffer>,
-    offer_data: Option<Rc<RefCell<OfferData>>>,
-    seat: Seat,
-    callback: Rc<RefCell<C>>,
+    offer_data: Option<Arc<Mutex<ServerDndOfferData>>>,
+    seat: Seat<D>,
 }
 
-impl<C: 'static> ServerDnDGrab<C> {
+impl<D> ServerDnDGrab<D> {
     pub(crate) fn new(
         start_data: PointerGrabStartData,
         metadata: super::SourceMetadata,
-        seat: Seat,
-        callback: Rc<RefCell<C>>,
-    ) -> ServerDnDGrab<C> {
-        ServerDnDGrab {
+        seat: Seat<D>,
+    ) -> Self {
+        Self {
             start_data,
             metadata,
             current_focus: None,
             pending_offers: Vec::with_capacity(1),
             offer_data: None,
             seat,
-            callback,
         }
     }
 }
 
-impl<C> PointerGrab for ServerDnDGrab<C>
+impl<D> PointerGrab<D> for ServerDnDGrab<D>
 where
-    C: FnMut(ServerDndEvent) + 'static,
+    D: DataDeviceHandler,
+    D: ServerDndGrabHandler,
```

Comment:
We could require `trait ServerDndGrabHandler: DataDeviceHandler` to remove the redundant trait requirements.

Same with the client.